ex1/dates_times.py

```diff
@@ -30,15 +30,60 @@
 Exercises for using the datetime and the calendar module
 '''
 
+#########################
+# Name: Patrick Hurnaus #
+# Matrikel-Nr: 1227116  #
+#########################
+
 # Define a function named last_of_month that takes an argument dt of type date
 # and returns a date object representing the last day of the month dt was in.
+import calendar
+from datetime import date, time, datetime
+
+# function "monthrange" is called, which returns a weekday and the number of days for the month
+# a new object date is created with the parameters and the and the last day of the month
+# (which is equal to the number of days)
+def last_of_month(dt):
+    mr = calendar.monthrange(dt.year, dt.month)
+    last = date(dt.year, dt.month, mr[1])
+    return last
+
 
 # Define a function named feed_the_gremlin which takes a time object as an
 # argument. It should return False between midnight and 6:30AM and True
 # otherwise.
 
+
+# two time objects are created, which represents midnight and morning
+# then it is checked if the time parameter t is in this interval
+def feed_the_gremlin(t):
+    midnight = time(0, 0, 0)
+    morning = time(6, 29, 59)
+    if midnight <= t <= morning:
+        return False
+    else:
+        return True
+
+
 # Define a function named how_long that takes two datetime objects dt and ref
 # where ref is the reference datetime, calculates the difference between them and
 # returns the difference as a string formatted like:
 # "01 days, 01 minutes, 01 seconds until 2000-12-31 15:59:59"
 # If ref is before dt then use 'since' instead of 'until'
+
+
+# first the difference between the two dates is calculated
+# total_secons() is called to get an integer value, which can be checked if it is positive or negative
+# if negative it means that the date dt is in the past
+# a string str_ouput is created with the format function which gives the parameter days, minutes and the ref date
+def how_long(dt, ref):
+    diff = ref - dt;
+    if diff.total_seconds() < 0:
+        diff = dt - ref
+        since_until = "since"
+    else:
+        since_until = "until"
+    str_output = ("{0} days, {1} minutes, {2} seconds " + since_until + " {3}").format(diff.days,
```

Comment:
It would be better to be explicit about the wanted format for the strings and date.